inference-engine/src/mkldnn_plugin/nodes/mkldnn_rnn.cpp

```diff
@@ -248,119 +278,149 @@ void MKLDNNRNN::fillSeqDesc() {
+        out_data_d[RNNInOutKind::CellState] = {S_4D_shape, memory::data_type::f32, memory::format_tag::ldnc};
+    }
 
-    if (bias)
-        w_bias_d = {{L, D, Gb, SC}, memory::data_type::f32, memory::format_tag::ldgo};
+    w_data_d  = {{L, D, DC, G, SC}, dataType, memory::format_tag::ldigo};
+    w_state_d = {{L, D, SC, G, SC}, dataType, memory::format_tag::ldigo};
 
-    // Try to create descriptor and corresponding configuration
-    in_data_d = {in_data_dims, memory::data_type::f32, memory::format_tag::tnc};
-    out_data_d = {out_data_dims, memory::data_type::f32, memory::format_tag::tnc};
+    if (bias && bias->size() != Gb * SC)
+        IE_THROW() << "RNN Layer. Biases size is not correct. Expected size:" << G * SC;
 
-    std::vector<TensorDesc> in_candidate;
-    if (nativeOrder)
-        in_candidate.push_back(in_data_d);
-    else
-        in_candidate.push_back(MKLDNNMemoryDesc{{N, T, DC}, memory::data_type::f32, memory::format_tag::ntc});
+    if (bias)
+        w_bias_d = {{L, D, Gb, SC}, memory::data_type::f32, memory::format_tag::ldgo};
 
-    for (int i = 1; i < ins.size(); i++)
-        in_candidate.emplace_back(MKLDNNMemoryDesc {S_shape, memory::data_type::f32, memory::format_tag::nc});
+    std::vector<TensorDesc> in_candidate, out_candidate;
 
-    std::vector<TensorDesc> out_candidate;
     if (nativeOrder) {
-        out_candidate.push_back(out_data_d);
+        in_candidate.push_back(in_data_d[RNNInOutKind::Layer]);
+        out_candidate.push_back(out_data_d[RNNInOutKind::Layer]);
     } else {
-        out_candidate.push_back(MKLDNNMemoryDesc{{N, T, SC}, memory::data_type::f32, memory::format_tag::ntc});
+        in_candidate.emplace_back(MKLDNNMemoryDesc{{N, T, DC}, dataType, memory::format_tag::ntc});
+        out_candidate.emplace_back(MKLDNNMemoryDesc{{N, T, SC}, dataType, memory::format_tag::ntc});
     }
 
-    for (int i = 1; i < outs.size(); i++) {
+    in_candidate.emplace_back(MKLDNNMemoryDesc{S_shape, dataType, memory::format_tag::nc});
+    out_candidate.emplace_back(MKLDNNMemoryDesc{S_shape, dataType, memory::format_tag::nc});
+
+    if (haveCellState(cell_type)) {
+        in_candidate.emplace_back(MKLDNNMemoryDesc{S_shape, memory::data_type::f32, memory::format_tag::nc});
         out_candidate.emplace_back(MKLDNNMemoryDesc{S_shape, memory::data_type::f32, memory::format_tag::nc});
     }
 
+    Precision weights_prec = as<MemoryBlob>(weights)->getTensorDesc().getPrecision();
+
+    if (!verifyWeightsPrecision(runtimePrecision, weights_prec)) {
+        if (runtimePrecision == Precision::BF16 && weights_prec == Precision::FP32)
+            convertWeightsBlobPrecision(weights_prec, weightsByLayerPrec.at(runtimePrecision));
+    }
+
     createDescriptor(in_candidate, out_candidate);
 }
 
+void MKLDNNRNN::convertWeightsBlobPrecision(const Precision cur_precision,
+                                            const Precision new_precision) {
```

Comment:
Given that the implementation itself can correctly convert only blobs with 4 byte data type to 2 blob with 2 byte data type it doesn't make any sense to pass precisions as parameters. Can we just have `convertWeightsBlobToBF16()` routine?

Reply:
@EgorDuplensky, you see, I can write replies.

Reply:
convertWeightsBlobToBF16 was the initial implementation, then I decided to make it more general.
But I agree it is probably better to keep it simple.
Done.